Feature/basic postgres

vedantroy opened the pull request:
This PR adds basic postgres integration. 
- Documentation for managing postgres
- A script for creating the correct tables and populating them with test data in postgres
- The route `/products` which returns a JSON array of the products.

Notes:
- [Knex.js](http://knexjs.org/) is used to generate SQL queries instead of using raw SQL. The benefits are:
    1. Automatically generates secure SQL that avoids SQL injection.
    2. Knex is (mostly) agnostic to the underlying database, so we can change the actual database without much effort.
    3. Knex is much easier than using the default `pg` module, which doesn't support promises and only allows raw SQL. Knex also has better integration with JS.

vedantroy committed:
Add guide for postgres

vedantroy committed:
Install node-postgres and knex

vedantroy committed:
Add script to manage postgres

vedantroy committed:
Add /products route

stevenlaitran commented on gifted_build/src/server/routes/test.js:
```diff
@@ -1,8 +1,31 @@
 const express = require('express');
 const router = express.Router();
 
+router.get('/products/', (req, res) => {
```
If we are using this file for the product routes, this should have a second identifier within the route, such as '/products/list_all' in order for use to possibly add more routes down the line.

Also we should rename this file from test.js to products.js

stevenlaitran commented on gifted_build/src/server/routes/test.js:
```diff
@@ -1,8 +1,31 @@
+const knex = require('knex')({
+    client: 'pg',
+    connection: {
```
rather than hardcoding the configs into the route, I would recommend elevating the knex connection into the server file and then reading the elements from a database.ini file stored at root, so we can pass the connection to the multiple routes and the code will be relatively the same when switching from dev to prod. we can set a prod_database.ini file too for production configs, but don't worry about for now

vedantroy replied:
I put the knex connection stuff in a `database.js` file. I left it as hardcoded for now because there is no production configuration.

stevenlaitran commented on gifted_build/postgres_manage.js:
```diff
@@ -0,0 +1,97 @@
+const { Client } = require('pg');
+const knex = require('knex')({
```
this connection is fine for the testing environment and for development purposes, but you should rename the file as this is really only meant for development purposes and setting up the production environment should be its own file. I realllly don't want any drop table commands to be near production.

stevenlaitran commented on gifted_build/postgres_manage.js:
```diff
@@ -0,0 +1,97 @@
+    ])
+}
+
+(async () => {
```
I don't think you need to async this btw or any of the other db commands

vedantroy replied:
All of the knex.js functions are async/return a `Promise`, so I think I have to use `async`/`await` or `.then`. While this doesn't seem too great, the lower-level node-postgres/`pg` library, forces the user to use callbacks, which imo, is even worse.

I'll see if there's a way to make the code more readable.

stevenlaitran commented on gifted_build/postgres_manage.js:
```diff
@@ -0,0 +1,97 @@
+        console.log(`Creating ${product_images_table} table`)
+        await knex.schema.createTable(product_images_table, table => {
+            table.string('SKU', 255).references(`${products_table}.SKU`)
+            table.binary('image')
```
We should probably have a optional reference to what color the image of the product refers to

stevenlaitran commented on gifted_build/postgres_manage.js:
```diff
@@ -0,0 +1,97 @@
+            table.decimal('price', 6, 3) //6 digits precision (total number of digits), 5 digits scale (# of digits to right of decimal point)
+            // The product description displayed in bold to the right of the product image
+            table.string('description', 511)
+            // The "conventional" method is to use a linking table
```
This definitely fine for our table as we they didn't specify about having multiple descriptions per item

vedantroy committed:
Remove unnecessary import

vedantroy committed:
postgres manager script is for dev only

vedantroy committed:
Rename route test to products

vedantroy committed:
Change endpoint for all products

vedantroy committed:
Put database config in separate file

vedantroy committed:
Add column that specifies what color the image of the product refers to

vedantroy commented:
Okay, I added a few updates.
- The knex configuration is moved into a separate `database.js` file.
- The postgres manager script has been renamed/modified, to indicate it is development only.
- The API endpoint for listing all products has been cleaned up/renamed.
- Added a column to the `product_images` table.

It should be good now. @stevenlaitran thoughts?

vedantroy committed:
Remove colors JSON array from products table

stevenlaitran commented on gifted_build/src/server/dev-server.js:
```diff
@@ -25,6 +25,6 @@ app.use('*', (req, res, next) => {
 app.listen(3000, () => console.log('dev app listening on port 3000'))
 
-server.use(require('./routes/test'));
+server.use(require('./routes/products'));
```
This is good, but be sure to change this in the production server as well

stevenlaitran commented on gifted_build/src/server/database.js:
```diff
@@ -0,0 +1,22 @@
+    knex = knexConstructor({
+        client: 'pg',
+        connection: {
+          user: 'build',
```
I don't know if this is a knex thing, but for most postgres installations, the default user is "postgres"

vedantroy replied:
Knex uses `pg` underneath, which is the default low-level Postgres driver for Node. I didn't use "postgres" as the user because it has lots of super-user privileges and because I couldn't figure out how to configure the password for it.

vedantroy committed:
Change route test to products in production server

stevenlaitran reviewed (approved):
Looks good LGTM!
